pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java

```diff
@@ -170,10 +171,10 @@ private String getMultiStageQueryResponse(String query, String queryOptions, Htt
       return QueryException.ACCESS_DENIED_ERROR.toString();
     }
 
-    // Get brokers, only DEFAULT tenant is supported for now.
-    // TODO: implement logic that only allows executing query where all accessed tables are within the same tenant.
-    List<String> instanceIds = new ArrayList<>(_pinotHelixResourceManager.getAllInstancesForBrokerTenant(
-        TagNameUtils.DEFAULT_TENANT_NAME));
+    SqlNodeAndOptions sqlNodeAndOptions = RequestUtils.parseQuery(query);
+    List<String> tableNames = CalciteSqlParser.extractTableNamesFromNode(sqlNodeAndOptions.getSqlNode());
+
+    List<String> instanceIds = getCommonInstances(tableNames);
```

Comment:
this change will probably be useless as 
1. the actual broker API only supports default tenant (hard-coded in many places)
2. can only get routing table for its responsible set of tables, but it is responsible for generating stage plan for the entire query.

Thus my suggestion is should we let's limit the scope to "all tables belong to the same tenant, not necessarily 'default'" first?
and try to get it right before we support random tenant combination?

Reply:
> the actual broker API only supports default tenant (hard-coded in many places)

You mean for multistage? We are actually running it in a non-default tenant internally. Queries work if you directly hit the broker that you know is in the tenant that you want.

Reply:
> Thus my suggestion is should we let's limit the scope to "all tables belong to the same tenant, not necessarily 'default'" first?
This PR takes care of this scenario specifically. At present, queries work fine when we hit brokers directly of a non-default tenant as @ankitsultana mentioned. We want to allow those queries to be pushed down to broker from controller as well.

Reply:
yeah that's probably b/c the broker you hit is responsible for only a set of tenants and that set of tenants are responsible for all the tables in the query. nevertheless it is not meant to be fully bug-free just b/c it can run. 

I can think of several scenarios that this could go wrong in other cluster settings. Let's aim at a single tenant and add tests for different type of configuration of 
1. broker tenant
2. server tenant (same, subset, superset of broker)
3. table (identical to broker, identical to server but subset of broker, identical to broker, subset of server) and mix and match multiple tables with the above configuration

Reply:
I think the issue here is that we are finding common instances for tables and in case of multiple tenants sharing few instances, this can lead to cross-tenant queries. Will use brokerTenant from tableConfig and see if all the tables are under the same tenant.